File: 015_3sum.py
class Solution:
    def threeSum(self, nums: list[int]) -> list[list[int]]:
        ans = []
        nums.sort()

        for i in range(len(nums)):
            if (i > 0 and nums[i] == nums[i-1]):
                continue

            l, r = i+1, len(nums)-1
            print(l, r)
            while l < r:
                temp = nums[i] + nums[l] + nums[r]
                if temp > 0:
                    r -= 1
                elif temp < 0:
                    l += 1
                else:
                    print("Here!")
                    ans.append([nums[i], nums[l], nums[r]])
                    l += 1
                    while (nums[l] == nums[l-1] and l < r):
                        l += 1

        return ans

File: 015_3sum.ts
function threeSum(arr: number[]): number[][] {
    const ans: number[][] = [];

    arr = arr.sort((a, b) => a - b);
    console.log(arr);

    for (let i = 0; i < arr.length; i++) {
        if (i >= 0 && arr[i - 1] === arr[i]) {
            continue;
        }
        let low = i + 1;
        let high = arr.length - 1;
        while (low < high) {
            const temp = arr[i] + arr[low] + arr[high];
            if (temp === 0) {
                ans.push([arr[i], arr[low], arr[high]]);
                low++;
                while (arr[low] === arr[low - 1] && low < high) {
                    low++;
                }
            } else if (temp > 0) {
                high--;
            } else {
                low++;
            }
        }
    }

    return ans;
}
